### ISSUE TITLE
Insert a layer taken from a donor into an alternate image

### ISSUE THREAD

**Issue by FearTheBadger** [moby/moby#43522]
**Description**

> * I am unsure if this is a desired outcome, a bug, or even a concern, but there is a way that you can take a layer from an alternate docker build image and insert it into the middle of an alternate image. This insertion will present as a layer in the receiving image just as if it had been included in the original Dockerfile.
>
> * I have only worked with exporting an image (`docker save`), modifying it, and then importing this new image (`docker load`). I have not dived deeper into all the potential variations that can be accomplished by this strategy.
>
> * To calm my immediate fears, I found that the modified image has a different `IMAGE ID` than the original image.

### How to accomplish this:

1. Setup environment

   * Create directory structure

     ```shell-session
     mkdir -p orig/image donor/image
     ```

2. Build a base image to start from

   * Create a new `$EDITOR orig/Dockerfile`

     ```Dockerfile
     FROM ubuntu:latest
     
     RUN apt update && apt install -y cowsay
     ```

   * Build the new image

     ```shell-session
     docker build -t test:orig orig/
     ```

   * Export image to local file

     ```shell-session
     docker save test:orig > test_orig.tar
     tar -xf test_orig.tar -C orig/image/
     ```

3. Create a second image based off of that original

   * Create a new `$EDITOR donor/Dockerfile`

     ```Dockerfile
     FROM test:orig
     
     RUN /usr/games/cowsay "I'm new here and should not exist" > /cowsay.txt
     ```

   * Build the new image

     ```shell-session
     docker build -t test:donor .
     ```

   * Export image to local file

     ```shell-session
     docker save test:donor > test_donor.tar
     tar xf test_donar.tar -C donor/image/
     ```

4. Find the information needed to insert the layers

   * Get `sha256` hash of `test:donor`

     ```shell-session
     docker inspect -f "{{ range .RootFS.Layers }}{{ println . }}{{end}}" test:donor
     ```

     * Record the last hash in the list

       <donor_hash> \= `sha256:258c504b1160276a3a4c……`

   * Find the folder hash of the donor layer in `donor/image/` and copy to the `orig/image`.

     ```shell-session
     cat donor/image/manifest.json
     ```

     * Record the last layer hash in the list, omitting the `/layer.tar`.

       <donor_folder_hash> \= `7a1942eded12b4142f27a4d8……`

     * Copy that folder over to the original directory

       ```shell-session
       cp -rp donor/image/<donor_folder_hash>/ orig/image/
       ```

   * Find the head of the original image

     ```shell-session
     cat orig/image/manifest.json
     ```

     * Record the last layer hash in the list, omitting the `/layer.tar`

       <orig_head_hash> \= `822ff5037aa……`

   * Find the parent of the head layer

     ```shell-session
     cat orig/image/<orig_head_hash>/json
     ```

     * Record the `parent:` value hash

       <orig_parent_hash> \= `ded12b4142f27a4d8b……`

5. Insert the new layer into the stack

   * Change directories to `orig/image`

     ```shell-session
     cd orig/image
     ```

   * Insert the folder hash into the manifest.json

     ```shell-session
     $EDITOR manifest.json 
     ```

     * Add `“<donor_folder_hash>/layer.tar”, ` to the **second to last** entry of the array.

   * Insert the new layer hash into the image json file. (This is the only json file in this base dir with a hash for a file name).

     ```shell-session
     $EDITOR <file_hash>.json
     ```

     * At the end of the line/file in the layers section add `”<donor_hash>”,` to the **second to last entry** of the `diff_ids` array.

   * Point the donor layer’s parent at the parent of the head layer.

     ```shell-session
     cd <donor_folder_hash>/
     cp -p ../<orig_parent_hash>/json .
     $EDITOR json
     ```

     * Insert a new `id:` before the existing `id`.

       `”id”:”<donor_folder_hash>”`

     * Change the now second `id:` to `parent:` then remove the now second `parent:` id and value if it exists.

   * Point the parent of head image to the new layer.

     ```shell-session
     cd ../<orig_head_hash>/
     $EDITOR json
     ```

     * Change the `parent` value to the `<donor_folder_hash>`

   * Go back a directory and create a new image tar

     ```shell-session
     cd ..
     tar -cf ../../test_new.tar .
     ```

6. Remove the old images and load the newly modified image

   ```shell-session
   docker rmi test:orig test:donor ubuntu:latest
   docker load < ../../test_new.tar
   ```

7. Validate that the image was changed.

   ```shell-session
   docker run -it --rm --entrypoint "/bin/bash" test:orig
   # cat /cowsay.txt
   ```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
$ docker version
Client: Docker Engine - Community
 Version:           20.10.14
 API version:       1.41
 Go version:        go1.16.15
 Git commit:        a224086
 Built:             Thu Mar 24 01:47:44 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.14
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.15
  Git commit:       87a90dc
  Built:            Thu Mar 24 01:46:10 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.5.11
  GitCommit:        3df54a852345ae127d1fa3092b95168e4a88e2f8
 runc:
  Version:          1.0.3
  GitCommit:        v1.0.3-0-gf46b6ba
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.1-docker)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 20.10.14
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3df54a852345ae127d1fa3092b95168e4a88e2f8
 runc version: v1.0.3-0-gf46b6ba
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.18.0-348.2.1.el8_5.x86_64
 Operating System: Red Hat Enterprise Linux 8.5 (Ootpa)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.46GiB
 Name: brock-dev.rhel8.dev
 ID: RAKV:JMV2:NL6A:RIUB:YUDB:Q3CY:ESAV:I6KI:JB3S:Q75H:UZFF:VA27
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by cpuguy83** [comment]
This seems like its working exactly as expected.
An image is defined by a manifest.
That manifest includes the digest of the image config and the digest of each layer, and then the config also includes the digest of each layer.

If you change the layers in the manifest you have changed the image.
The Dockerfile does not matter and is meaingless except as a means of assmebling filesystem diffs and writing json.

Basically, you did what a Dockerfile does just with manual steps.

**Comment by cpuguy83** [comment]
Closing since this seems to be operating as expected, but feel free to discuss.
Thank you for reporting your concerns.
